### PR TITLE
Stock and flow variables on Godley icons should not be copied as groups.

### PR DESCRIPTION
**Pull request by wdednam**
This change is [Reviewable](https://reviewable.io/reviews/highperformancecoder/minsky/121)

### CODE REVIEW

File: model/canvas.cc [hunk 1]
```diff
@@ -629,7 +629,6 @@ namespace minsky
   void Canvas::copyVars(const std::vector<VariablePtr>& v)
   {
     auto group=model->addGroup(new Group);
-    setItemFocus(group);
```
Comment (highperformancecoder) [inline model/canvas.cc]: I suspect this code can be refactored to not create a new group. Unless we desire to share code between this and paste.

**Comment by highperformancecoder** [comment]
Need to fix the regression test.

File: model/canvas.cc [hunk 2]
```diff
@@ -628,44 +628,40 @@ namespace minsky
...
+          heights.push_back(rv.height());
+          maxWidth=max(maxWidth, widths.back());
+          totalHeight+=heights.back();
+        }
+      float y=v[0]->y() - totalHeight;
+      for (size_t i=0; i<v.size(); ++i)
+        {
+          auto ni=v[i]->clone();
+          ni->rotation=0;
+          //ni->zoomFactor=group->zoomFactor;
```
Comment (highperformancecoder) [inline model/canvas.cc]: dead code, should be removed

File: model/canvas.cc [hunk 3]
```diff
@@ -628,44 +628,40 @@ namespace minsky
...
+          ni->moveTo(v[0]->x()+maxWidth-v[i]->zoomFactor()*widths[i],
+                     y+heights[i]);
+          // variables need to refer to outer scope
+          if (ni->name()[0]!=':')
+            ni->name(':'+ni->name());
+          y+=2*v[i]->zoomFactor()*heights[i];
+          // Stock and flow variables on Godley icons should not be copied as groups. For ticket 1039
+          ItemPtr newI(ni);
```
Comment (highperformancecoder) [inline model/canvas.cc]: Not actually necessary to create the shared_ptr here.

File: model/canvas.cc [hunk 4]
```diff
@@ -628,44 +628,40 @@ namespace minsky
...
+          selection.insertItem(model->addItem(newI));		 
+        }
+        // Item focus put on one of the copied vars that are still in selection. For ticket 1039
+        setItemFocus(model->findAny
```
Comment (highperformancecoder) [inline model/canvas.cc]: What about using selection.items[0]? Obviously check for empty selection first, and clear the itemFocus in that case.

File: test/testModel.cc [hunk 5]
```diff
@@ -776,32 +776,15 @@ SUITE(Canvas)
         godley->table.cell(2,2)="y";
         godley->update();
 
-        
+        // Unit test greatly simplified since flow/stock vars no longer copied via temporary group. For ticket 1039
```
Comment (highperformancecoder) [inline test/testModel.cc]: This comment is not really helpful for down the track. Add it to the ticket discussion if you want to comment on it.

File: test/testModel.cc [hunk 6]
```diff
@@ -776,32 +776,15 @@ SUITE(Canvas)
...
         unsigned originalNumItems=model->numItems();
-        unsigned originalNumGroups=model->numGroups();
         copyAllFlowVars();
         CHECK_EQUAL(originalNumItems+godley->flowVars().size(),model->numItems());
```
Comment (highperformancecoder) [inline test/testModel.cc]: This is a good check. But should also check that the relevant items have been inserted into selection, and also that  (I'm guessing in this case) that there are now two copies of the flowVars in model. The way I would approach this is do a loop of model->items, and increment a map<string,int> where the key is the valueId, then afterwards, loop over the flowVars and check that the count is 2. eg

map<string,int> idCnt;
for (auto& i: model->items)
  if (auto v=i.variableCast())
     idCnt[v->valueId()]++;
for (auto v: godley->flowVars())
    CHECK_EQUAL(2, idCnt[v->valueId()])

File: model/canvas.cc (hunk 2 above)
Comment (wdednam) [reply model/canvas.cc]: Thanks. I've removed it.

File: model/canvas.cc (hunk 3 above)
Comment (wdednam) [reply model/canvas.cc]: Also fixed this. I've made ni an ItemPtr directly as you suggested, and to access members of variablePtr I use variableCast(), e.g., (ni->variableCast())->name().

File: model/canvas.cc (hunk 4 above)
Comment (wdednam) [reply model/canvas.cc]: Thanks, done. I forgot about that member of selection.

File: test/testModel.cc (hunk 5 above)
Comment (wdednam) [reply test/testModel.cc]: I've removed it, thanks.

File: test/testModel.cc (hunk 6 above)
Comment (wdednam) [reply test/testModel.cc]: Thank you for this, the test is much more robust now. I will soon commit the new changes.

File: model/canvas.cc [hunk 7]
```diff
@@ -628,32 +628,37 @@ namespace minsky
...
+                     y+heights[i]);
+          // variables need to refer to outer scope
+          if ((ni->variableCast())->name()[0]!=':')
+            (ni->variableCast())->name(':'+(ni->variableCast())->name());
+          y+=2*v[i]->zoomFactor()*heights[i];
+          selection.insertItem(model->addItem(ni));		 
+        }
+        // Item focus put on one of the copied vars that are still in selection. For ticket 1039
+        if (!selection.empty()) setItemFocus(selection.items[0]);
+        else selection.clear();
```
Comment (highperformancecoder) [inline model/canvas.cc]: Actually, selection is empty, so doesn't need to be cleared. However, itemFocus should probably be cleared too (setItemFocus(nullptr) perhaps?).

Comment (wdednam) [reply model/canvas.cc]: Thanks, I will fix this after your current commit finishes and I can merge the most up to date version into this branch.

Comment (highperformancecoder) [reply model/canvas.cc]: OK - I've finished with #1113 now. There won't be any merge conflicts with this PR.

Comment (wdednam) [reply model/canvas.cc]: I don't know why test t0001a.sh fails, it doesn't on my machine.

File: model/canvas.cc [hunk 8]
```diff
@@ -677,32 +677,37 @@ namespace minsky
...
+          widths.push_back(rv.width());
+          heights.push_back(rv.height());
+          maxWidth=max(maxWidth, widths.back());
+          totalHeight+=heights.back();
+        }
+      float y=v[0]->y() - totalHeight;
+      for (size_t i=0; i<v.size(); ++i)
+        {
+		  // Stock and flow variables on Godley icons should not be copied as groups. For ticket 1039	
+          ItemPtr ni(v[i]->clone());
```
Comment (highperformancecoder) [inline model/canvas.cc]: I would add an assert(ni->variableCast()) here as a protection. Yes the current code guarantees that ni is actually a VariableBase, but things have a habit of changing :).

Also variableCast() is not without cost. It is cheaper than dynamic_cast. In this case, you could use static_cast, which is free. Mind you, this is not really performance critical code.

**Pull request merged**